pkg/vsphere/spec/spec.go

```diff
@@ -248,3 +251,42 @@ func (s *VirtualMachineConfigSpec) generateNextKey() int32 {
 	s.key -= 10
 	return s.key
 }
+
+// Spec returns the base types.VirtualMachineConfigSpec object
+func (s *VirtualMachineConfigSpec) Spec() *types.VirtualMachineConfigSpec {
+	return s.VirtualMachineConfigSpec
+}
+
+// VirtualDeviceSlotNumber returns the PCI slot number of a device
+func VirtualDeviceSlotNumber(d types.BaseVirtualDevice) int32 {
+	s := d.GetVirtualDevice().SlotInfo
+	if s == nil {
+		return NilSlot
+	}
+
+	if i, ok := s.(*types.VirtualDevicePciBusSlotInfo); ok {
+		return i.PciSlotNumber
+	}
+
+	return NilSlot
+}
+
+// CollectSlotNumbers returns a collection of all the PCI slot numbers for devices in the spec
+func (s *VirtualMachineConfigSpec) CollectSlotNumbers() []int32 {
+	// collect all the already assigned slot numbers
+	var slots map[int32]bool
```

Comment:
I mainly had it in there for slot number duplicates. The code is getting exercised, and the reason it doesn't panic is because it never gets to line 280. 

It doesn't really make sense the way the map is used here, since it is not really detecting duplicates. I am going to take it and assume there are no duplicate slot numbers in the spec.